runtime/commonMain/src/kotlinx/serialization/SealedSerializer.kt

```diff
@@ -75,26 +74,45 @@ public class SealedClassSerializer<T : Any>(
     override val descriptor: SerialDescriptor =
         SealedClassDescriptor(serialName, subclassSerializers.map { it.descriptor })
 
-    private val backingMap: Map<KClass<out T>, KSerializer<out T>>
-    private val inverseMap: Map<String, KSerializer<out T>>
+    private val class2Serializer: Map<KClass<out T>, KSerializer<out T>>
+    private val serialName2Serializer: Map<String, KSerializer<out T>>
+    private val serializer2AllProperties: Map<KSerializer<*>, Set<String>>
 
     init {
         require(subclasses.size == subclassSerializers.size) {
             "Arrays of classes and serializers must have the same length," +
                     " got arrays: ${subclasses.contentToString()}, ${subclassSerializers.contentToString()}"
         }
-        backingMap = subclasses.zip(subclassSerializers).toMap()
-        inverseMap = backingMap.values.associateBy { serializer -> serializer.descriptor.serialName }
+        class2Serializer = subclasses.zip(subclassSerializers).toMap()
+        serialName2Serializer = class2Serializer.entries.groupingBy { it.value.descriptor.serialName }
+            .aggregate<Map.Entry<KClass<out T>, KSerializer<out T>>, String, Map.Entry<KClass<*>, KSerializer<out T>>>
+            { key, accumulator, element, _ ->
+                if (accumulator != null) {
+                    error("Multiple sealed subclasses of $baseClass have the same serial name '$key: ${accumulator.key}, ${element.key}")
+                }
+                element
+            }.mapValues { it.value.value }
+        serializer2AllProperties = subclassSerializers.associate { it to it.descriptor.elementNames().toSet() }
     }
 
     override fun findPolymorphicSerializer(decoder: CompositeDecoder, klassName: String): KSerializer<out T> {
-        return inverseMap[klassName]
-                ?: super.findPolymorphicSerializer(decoder, klassName)
+        return serialName2Serializer[klassName] ?: super.findPolymorphicSerializer(decoder, klassName)
     }
 
     override fun findPolymorphicSerializer(encoder: Encoder, value: T): KSerializer<out T> {
-        return backingMap[value::class]
-                ?: super.findPolymorphicSerializer(encoder, value)
+        return class2Serializer[value::class] ?: super.findPolymorphicSerializer(encoder, value)
+    }
+
+    internal fun validate(actualSerializer: KSerializer<Any>, classDiscriminator: String) {
+        val properties = serializer2AllProperties[actualSerializer] ?: return
+        if (classDiscriminator !in properties) return
+        val clazz = class2Serializer.entries.first { it.value === actualSerializer }.key
+        error(
+            "Sealed $clazz cannot be serialized as base class $baseClass because it has property name that conflicts with JSON class discriminator. " +
+                    "You can either change class discriminator in JsonConfiguration, " +
```

Comment:
Don't refer to JSON as it is/should be format agnostic. If this is a JSON specific issue then it should be somehow handled in the format, not in the agnostic code.

Reply:
This exact method is JSON-specific (and internal anyway).
In order to make it format-agnostic, we should expose backing maps (with properties and KClasses) to public and check them from within in JSON.
I don't mind doing it, but only when we have a compelling use-case to make it public.